### PR TITLE
Adds include and exclude at match to Lambda log forwarder

### PR DESCRIPTION
**Pull request by le-he-hoo**
### What does this PR do?

Adds `INCLUDE_AT_MATCH` and `EXCLUDE_AT_MATCH` to Lambda log forwarder
### Motivation

Customer  request

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

**Comment by le-he-hoo** [comment]
TODO: update README.md

File: aws/logs_monitoring/lambda_function.py [hunk 1]
```diff
@@ -465,6 +486,48 @@ def split(events):
     return metrics, logs
 
 
+# should only be called when INCLUDE_AT_MATCH and/or EXCLUDE_AT_MATCH exist
+def filter_logs(logs):
```
Comment (ajacquemot) [inline aws/logs_monitoring/lambda_function.py:488]: I think you could simplify this method by iterating over all logs and for each log, try include, try exclude, this would limit the cases to two.

Comment (ajacquemot) [reply aws/logs_monitoring/lambda_function.py:488]: The question now is which rule wins where the log should both included and excluded, on the agent we decided to exclude it https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/processor/processor.go#L78

File: aws/logs_monitoring/lambda_function.py [hunk 2]
```diff
@@ -80,6 +80,27 @@ def __init__(self, name, pattern, placeholder):
     )
 ]
 
+
+# Filtering logs 
+# Option to include or exclude logs based on a pattern match
+INCLUDE_AT_MATCH = os.getenv("INCLUDE_AT_MATCH", default=None)
+if INCLUDE_AT_MATCH is not None:
```
Comment (ajacquemot) [inline aws/logs_monitoring/lambda_function.py]: We can see code duplication here, the exact same logic is also used inside the scrubber, I think we should isolate this logic inside a method, and use from different places.
The interface would look like:
```
def compileRegex(pattern):
  # TODO: implement me
  pass
```

**Review by ajacquemot — COMMENTED** [review]
Thanks for working on this, the logic is here, let's factorise and simplify a bit what we have 👍

File: aws/logs_monitoring/lambda_function.py (hunk 1 above)
Comment (le-he-hoo) [reply aws/logs_monitoring/lambda_function.py:488]: Exclude 🛑 wins here too 👍

Simplified it a bit.

File: aws/logs_monitoring/lambda_function.py [hunk 3]
```diff
@@ -465,6 +486,34 @@ def split(events):
...
+    If no filtering rules exist, return all the logs.
+    """
+    if INCLUDE_AT_MATCH is None and EXCLUDE_AT_MATCH is None:
+        return logs 
+    # Add logs that should be sent to logs_to_send
+    logs_to_send = []
+    # Test each log for exclusion and inclusion, if the criteria exist
+    for log in logs:
+        try:
+            string_log = json.dumps(log)
```
Comment (ajacquemot) [inline aws/logs_monitoring/lambda_function.py]: Here we are transforming logs to json-string and storing the logs at the end, could we store a json-string instead and change the logic downstream to manipulation string instead of objects so that we don't do the object-to-string operation twice ?

File: aws/logs_monitoring/lambda_function.py [hunk 4]
```diff
@@ -74,12 +74,33 @@ def __init__(self, name, pattern, placeholder):
         "xxxxx@xxxxx.com",
     ),
     ScrubbingRuleConfig(
-        "CUSTOM_SCRUBBING_RULE", 
-        os.getenv("CUSTOM_SCRUBBING_RULE", default=None), 
-        os.getenv("CUSTOM_SCRUBBING_RULE_REPLACEMENT", default="xxxxx")
+        "DD_SCRUBBING_RULE",
```
Comment (ajacquemot) [inline aws/logs_monitoring/lambda_function.py:77]: The change on scrubbing rules could have been done in another PR to ease review

File: aws/logs_monitoring/lambda_function.py (hunk 2 above)
Comment (ajacquemot) [reply aws/logs_monitoring/lambda_function.py]: I can help with this one if you want, feel free to ping me IRC.

File: aws/logs_monitoring/lambda_function.py (hunk 4 above)
Comment (le-he-hoo) [reply aws/logs_monitoring/lambda_function.py:77]: Good point!

File: aws/logs_monitoring/lambda_function.py (hunk 3 above)
Comment (le-he-hoo) [reply aws/logs_monitoring/lambda_function.py]: I went for it and made a few downstream changes.

File: aws/logs_monitoring/lambda_function.py (hunk 2 above)
Comment (le-he-hoo) [reply aws/logs_monitoring/lambda_function.py]: I added a `compileRegex` and used it for the include, exclude, and scrubbing rules.

File: aws/logs_monitoring/lambda_function.py [hunk 5]
```diff
@@ -465,6 +484,35 @@ def split(events):
     return metrics, logs
 
 
+# should only be called when INCLUDE_AT_MATCH and/or EXCLUDE_AT_MATCH exist
+def filter_logs(logs):
```
Comment (ajacquemot) [inline aws/logs_monitoring/lambda_function.py:488]: would be nice that this method does not have to transform the logs to json, can we do it upstream please ?

File: aws/logs_monitoring/lambda_function.py [hunk 6]
```diff
@@ -251,7 +272,7 @@ def send(self, logs):
         try:
             resp = self._session.post(
                 self._url,
-                data=self._scrubber.scrub(json.dumps(logs)),
+                data=self._scrubber.scrub("[{}]".format(", ".join(map(json.dumps, logs)))),
```
Comment (ajacquemot) [inline aws/logs_monitoring/lambda_function.py]: Why do we need to `json.dumps` again ? We already do it upstream right ? I think we just need to join the jsonString with `,` and wrap this under `[]`

Comment (le-he-hoo) [reply aws/logs_monitoring/lambda_function.py]: You're right.  Good catch.

File: aws/logs_monitoring/lambda_function.py (hunk 5 above)
Comment (le-he-hoo) [reply aws/logs_monitoring/lambda_function.py:488]: @ajacquemot Sure! Where do you think is a good place upstream to do it?

Comment (le-he-hoo) [reply aws/logs_monitoring/lambda_function.py:488]: Moved it to the `split` method

**Review by ajacquemot — APPROVED** [review]
LGTM, I think we should test thoroughly this branch before merging as we touched the frame for TCP and the payload for HTTP as well as the regexes. Once you validated that nothing broke, let's merge it 👍
